src/array.c

```diff
@@ -485,8 +499,15 @@ JL_DLLEXPORT jl_value_t *jl_arrayref(jl_array_t *a, size_t i)
     assert(i < jl_array_len(a));
     jl_value_t *elt;
     if (!a->flags.ptrarray) {
-        jl_value_t *el_type = (jl_value_t*)jl_tparam0(jl_typeof(a));
-        elt = jl_new_bits(el_type, &((char*)a->data)[i*a->elsize]);
+        jl_value_t *eltype = (jl_value_t*)jl_tparam0(jl_typeof(a));
+        if (jl_is_uniontype(eltype)) {
+            // isbits union selector bytes are always stored directly after the last array element
+            uint8_t sel = ((uint8_t*)a->data)[jl_array_len(a) * a->elsize + i];
```

Comment:
I started doing it that way, but it seemed to make codgen and mutating operations a lot hairier. I don't quite remember the specifics, but I finally switched over to right after the last element and it seemed overall simpler.